File: app/src/main/java/com/icerrate/popularmovies/view/movies/detail/TrailersAdapter.java
package com.icerrate.popularmovies.view.movies.detail;

import android.content.Context;
import android.support.v7.widget.AppCompatImageButton;
import android.support.v7.widget.AppCompatImageView;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;

import com.icerrate.popularmovies.R;
import com.icerrate.popularmovies.data.model.Trailer;
import com.icerrate.popularmovies.view.common.GlideApp;

import java.util.ArrayList;
import java.util.List;

/**
 * @author Ivan Cerrate.
 */

public class TrailersAdapter extends RecyclerView.Adapter<TrailersAdapter.TrailerViewHolder> {

    private List<Trailer> items;

    private OnItemClickListener onItemClickListener;

    public TrailersAdapter(OnItemClickListener onItemClickListener) {
        this(new ArrayList<Trailer>(), onItemClickListener);
    }

    public TrailersAdapter(List<Trailer> items, OnItemClickListener onItemClickListener) {
        this.items = items;
        this.onItemClickListener = onItemClickListener;
    }

    @Override public TrailerViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_trailer, parent, false);
        return new TrailerViewHolder(view, onItemClickListener);
    }

    @Override public void onBindViewHolder(TrailerViewHolder viewHolder, int position) {
        Trailer trailer = items.get(position);

        Context context = viewHolder.itemView.getContext();
        String url = trailer.getVideoThumbnail();
        GlideApp.with(context)
                .load(url)
                .placeholder(context.getResources().getDrawable(R.drawable.backdrop_placeholder))
                .into(viewHolder.thumbnailImageView);
        viewHolder.itemView.setTag(trailer);
        viewHolder.playImageButton.setTag(trailer);
    }

    @Override public int getItemCount() {
        return items.size();
    }

    public void addItems(List<Trailer> items) {
        this.items.addAll(items);
        notifyDataSetChanged();
    }

    public static class TrailerViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        public AppCompatImageView thumbnailImageView;

        public AppCompatImageButton playImageButton;

        private OnItemClickListener onItemClickListener;

        public TrailerViewHolder(View itemView, OnItemClickListener onItemClickListener) {
            super(itemView);
            thumbnailImageView = itemView.findViewById(R.id.thumbnail);
            playImageButton = itemView.findViewById(R.id.play);
            this.onItemClickListener = onItemClickListener;
            itemView.setOnClickListener(this);
            playImageButton.setOnClickListener(this);
        }

        @Override
        public void onClick(View view) {
            if (onItemClickListener != null) {
                onItemClickListener.onItemClick(view);
            }
        }
    }

    public interface OnItemClickListener {

        void onItemClick(View view);
    }
}

File: app/src/main/java/com/icerrate/popularmovies/view/movies/detail/TrailersAdapter.kt
package com.icerrate.popularmovies.view.movies.detail

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.icerrate.popularmovies.R
import com.icerrate.popularmovies.data.model.Trailer
import com.icerrate.popularmovies.databinding.ItemTrailerBinding

/**
 * @author Ivan Cerrate.
 */
class TrailersAdapter(
    private val items: MutableList<Trailer> = mutableListOf(),
    private val onItemClickListener: OnItemClickListener?
) : RecyclerView.Adapter<TrailersAdapter.TrailerViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): TrailerViewHolder {
        val binding = ItemTrailerBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return TrailerViewHolder(binding, onItemClickListener)
    }

    override fun onBindViewHolder(viewHolder: TrailerViewHolder, position: Int) {
        val trailer = items[position]
        viewHolder.bind(trailer)
    }

    override fun getItemCount(): Int = items.size

    fun addItems(newItems: List<Trailer>) {
        this.items.addAll(newItems)
        notifyDataSetChanged()
    }

    class TrailerViewHolder(
        private val binding: ItemTrailerBinding,
        private val onItemClickListener: OnItemClickListener?
    ) : RecyclerView.ViewHolder(binding.root), View.OnClickListener {

        init {
            itemView.setOnClickListener(this)
            binding.play.setOnClickListener(this)
        }

        fun bind(trailer: Trailer) {
            val context = itemView.context
            val url = trailer.getVideoThumbnail()

            Glide.with(context)
                .load(url)
                .placeholder(ContextCompat.getDrawable(context, R.drawable.backdrop_placeholder))
                .into(binding.thumbnail)

            itemView.tag = trailer
            binding.play.tag = trailer
        }

        override fun onClick(view: View) {
            onItemClickListener?.onItemClick(view)
        }
    }

    interface OnItemClickListener {
        fun onItemClick(view: View)
    }
}